provider/etcd/etcd_client.go

```diff
@@ -0,0 +1,337 @@
+  As other Alertmanager in the Alertmanager cluster write to Etcd, all of the
+  Alertmanagers receives those updates.
+
+* For the last part of initialization, Etcd will proceed to read all alert
+  objects from Etcd.
+
+Ongoing Operation (Async)
+
+* Alertmanager can receive alerts from both API requests AND the Etcd watch
+  subscription.  Alerts from both of these sources are sent to the etcd alert
+  provider via Put().  The alert is put into Etcd only when there is a
+  difference between the Etcd alert and memory.
+
+* The Etcd clientv3 library will automatically reconnect if the Etcd server
+  goes down or up, or upon resuming network connectivity if it had been lost.
+
+* The long running watch subscription will reconnect if the watch channel is
+  closed.
+
+*/
+
+package etcd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	"go.etcd.io/etcd/clientv3"
+	"google.golang.org/grpc"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/alertmanager/types"
+)
+
+var (
+	ErrorEtcdNotInitialized = errors.New("Etcd not initialized")
+)
+
+type EtcdClient struct {
+	alerts    *Alerts
+	endpoints []string
+	prefix    string
+	logger    log.Logger
+	client    *clientv3.Client
+	mtx       sync.Mutex
+}
+
+func NewEtcdClient(ctx context.Context, a *Alerts, endpoints []string, prefix string) (*EtcdClient, error) {
+
+	ec := &EtcdClient{
+		alerts:    a,
+		endpoints: endpoints,
+		prefix:    prefix,
+		logger:    log.With(a.logger, "component", "provider.etcd"),
+	}
+
+	// create the configuration
+	etcdConfig := clientv3.Config{
+		Endpoints:        endpoints,
+		AutoSyncInterval: 60 * time.Second,
+		DialTimeout:      10 * time.Second,
+		DialOptions:      []grpc.DialOption{grpc.WithBlock()}, // block until connect
+	}
+
+	// create the client
+	client, err := clientv3.New(etcdConfig)
+	if err != nil {
+		// On startup, if we cannot connect to the etcd cluster, then fail hard so that the
+		// user may address a potential configuration issue.  Once the clientv3 connects
+		// successfully, clientv3 will reconnect to the etcd cluster as it goes down or up,
+		// or into or out of network connectivity.
+		level.Error(ec.logger).Log("msg", "Etcd connection failed", "err", err)
+		os.Exit(1)
+	} else {
+		level.Info(ec.logger).Log("msg", "Etcd connection successful")
+	}
+	ec.mtx.Lock()
+	ec.client = client
+	ec.mtx.Unlock()
+
+	// start a goroutine to ensure the client will be cleaned up when the context is done
+	go func() {
+		defer func() {
+			ec.mtx.Lock()
+			if ec.client != nil {
+				ec.client.Close()
+				ec.client = nil
+			}
+			ec.mtx.Unlock()
+			level.Info(ec.logger).Log("msg", "Etcd connection shut down")
+		}()
+
+		for range ctx.Done() {
+		}
+	}()
+	return ec, nil
+}
+
+func (ec *EtcdClient) CheckAndPut(alert *types.Alert) error {
```

Comment:
FYI: Functions missing proper top of function comments. This may reflect poorly in godoc. You can check that locally: https://godoc.org/golang.org/x/tools/cmd/godoc